File: scripts/moveit_ik_demo.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

# Copyright 2020 The UmbraTek Inc. All Rights Reserved.
#
# Software License Agreement (BSD License)
#
# Author: johnson huang<johnson@umbratek.com>
# =============================================================================

import rospy, sys
import moveit_commander
from geometry_msgs.msg import PoseStamped, Pose
#api reference https://github.com/ros-planning/moveit/blob/master/moveit_commander/src/moveit_commander/move_group.py


class MoveItIkDemo:
    def __init__(self):
        moveit_commander.roscpp_initialize(sys.argv)
        rospy.init_node('utra_moveit_ik_demo')
        arm = moveit_commander.MoveGroupCommander('manipulator')
                
        end_effector_link = arm.get_end_effector_link()
                        
        # set pose reference Coordinate System
        reference_frame = 'base_link'
        arm.set_pose_reference_frame(reference_frame)
        
        arm.allow_replanning(True)
        
        # set tolerance
        arm.set_goal_position_tolerance(0.001) # meter unit
        arm.set_goal_orientation_tolerance(0.01) # rad unit
       
        # set max velocity acceleration
        arm.set_max_acceleration_scaling_factor(0.1)
        arm.set_max_velocity_scaling_factor(0.1)

        # go to the init pose
        arm.set_named_target('init')
        arm.go()
        rospy.sleep(3)

        # get current pose     
        current_pose = arm.get_current_pose()
        print(current_pose)
        
        # set target pose
        target_pose = PoseStamped()
        target_pose.header.frame_id = reference_frame
        target_pose.header.stamp = rospy.Time.now()    
        target_pose.pose.position.x = current_pose.pose.position.x
        target_pose.pose.position.y = current_pose.pose.position.y
        target_pose.pose.position.z = current_pose.pose.position.z-0.1 # move down 0.1m
        target_pose.pose.orientation.x = current_pose.pose.orientation.x
        target_pose.pose.orientation.y = current_pose.pose.orientation.y
        target_pose.pose.orientation.z = current_pose.pose.orientation.z
        target_pose.pose.orientation.w = current_pose.pose.orientation.w
        
      
        arm.set_start_state_to_current_state()
        
        # set to target pose and move
        arm.set_pose_target(target_pose,end_effector_link)
        traj = arm.plan()
        arm.execute(traj)
        rospy.sleep(1)

      
        moveit_commander.roscpp_shutdown()
        moveit_commander.os._exit(0)

if __name__ == "__main__":
    MoveItIkDemo()

    
    

File: src/moveit_ik_demo.cpp
/***********************************************************************
# Copyright 2020 The UmbraTek Inc. All Rights Reserved.
#
# Software License Agreement (BSD License)
#
# Author: johnson huang<johnson@umbratek.com>
# 
***********************************************************************/

#include <string>
#include <ros/ros.h>
#include <moveit/move_group_interface/move_group_interface.h>

int main(int argc, char **argv)
{
    ros::init(argc, argv, "utra_moveit_fk_demo");
    ros::AsyncSpinner spinner(1);
    spinner.start();

    moveit::planning_interface::MoveGroupInterface arm("manipulator");

    std::string end_effector_link = arm.getEndEffectorLink();

    // set pose reference Coordinate System
    std::string reference_frame = "base_link";
    arm.setPoseReferenceFrame(reference_frame);

    arm.allowReplanning(true);

    // set tolerance
    arm.setGoalPositionTolerance(0.001); // meter unit
    arm.setGoalOrientationTolerance(0.01); // rad unit

    // set max velocity acceleration
    arm.setMaxAccelerationScalingFactor(0.1);
    arm.setMaxVelocityScalingFactor(0.1);

    // go to the init pose
    arm.setNamedTarget("init");
    arm.move();
    sleep(3);

    // get current pose     
    geometry_msgs::PoseStamped current_pose = arm.getCurrentPose();

    // set target pose
    geometry_msgs::Pose target_pose;
    target_pose.orientation.x = current_pose.pose.orientation.x;
    target_pose.orientation.y = current_pose.pose.orientation.y;
    target_pose.orientation.z = current_pose.pose.orientation.z;
    target_pose.orientation.w = current_pose.pose.orientation.w;

    target_pose.position.x = current_pose.pose.position.x;
    target_pose.position.y = current_pose.pose.position.y;
    target_pose.position.z = current_pose.pose.position.z-0.1; // move down 0.1m

    arm.setStartStateToCurrentState();
    //set to target pose and move
    arm.setPoseTarget(target_pose);

    // start plan first
    moveit::planning_interface::MoveGroupInterface::Plan plan;
    moveit::planning_interface::MoveItErrorCode success = arm.plan(plan);

    ROS_INFO("Plan (pose goal) %s",success?"":"FAILED");   

    //execute the plan
    if(success)
      arm.execute(plan);
    sleep(1);


    ros::shutdown(); 

    return 0;
}
